block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockMetadataStore.scala

```diff
@@ -130,32 +162,56 @@ object BlockMetadataStore {
       hash: BlockHash,
       parents: Set[BlockHash],
       blockNum: Long,
-      isInvalid: Boolean
+      isInvalid: Boolean,
+      isDirectlyFinalised: Boolean
   )
 
   private def recreateInMemoryState(
-      blocksInfoMap: Map[BlockHash, BlockInfo],
-      lastFinalizedBlockHash: Option[BlockHash]
+      blocksInfoMap: Map[BlockHash, BlockInfo]
   ): DagState = {
-    val emptyState =
-      DagState(dagSet = Set(), childMap = Map(), heightMap = SortedMap(), finalizedBlockSet = Set())
+    val emptyState: DagState =
+      DagState(
+        dagSet = Set(),
+        childMap = Map(),
+        heightMap = SortedMap(),
+        lastFinalizedBlock = none[(BlockHash, Long)],
+        finalizedBlockSet = Set()
+      )
 
     // Add blocks to DAG state
     val dagState = blocksInfoMap.foldLeft(emptyState) {
       case (state, (_, block)) => addBlockToDagState(block)(state)
     }
 
-    // Calculate set of finalized blocks
-    val finalizedBlockSet = lastFinalizedBlockHash.fold(Set.empty[BlockHash])(lbh => {
-      DagOps
-        .bfTraverse(List(lbh))(
-          bh => blocksInfoMap.get(bh).map(_.parents.toList).getOrElse(List.empty)
+    // Find last finalized block (if any). Traverse from latest height down, return first directly finalized.
+    val lfbOpt = dagState.heightMap.keys.toList.reverse.iterator
+      .map(h => dagState.heightMap(h).find(blocksInfoMap(_).isDirectlyFinalised))
+      .collectFirst { case Some(dfb) => dfb }
+      .map { lfb =>
+        (
+          lfb,
+          dagState.heightMap
+            .find { case (_, v) => v.contains(lfb) }
+            .map {
+              case (height, _) => height
+            }
+            .get
         )
-        .toSet
-    })
+      }
+
+    // Calculate set of finalized blocks
+    val finalizedBlockSet = lfbOpt.fold(Set.empty[BlockHash])(
+      lfb =>
+        DagOps
+          .bfTraverse(List(lfb._1))(
+            bh => blocksInfoMap.get(bh).map(_.parents.toList).getOrElse(List.empty)
+          )
+          .toSet
+    )
```

Comment:
This seems to complicated. Try to add forth _update_ to `addBlockToDagState` for LFB.

Reply:
I did not change logic here, don't want touch too much. You cannot get finalized set until you get LFB. So you need to push everything in `addBlockToDagState`, which to me looks like a complication.

Reply:
BlockMetaData, BlockMetaDataStore, in memory representation are different levels of indexing what ever is stored in _raw_ blocks. So we need to design this index in a way that simplify our code and increase performance.

With that in mind, we can put in BlockMetaData what we need. E.g. we can add both flags for directly and indirectly finalized blocks which will reduce amount of traversals and creation of in memory index.

Reply:
Implemented in https://github.com/rchain/rchain/pull/3459/commits/208c4fe33d5963252631f1902b4dd9c58455acda